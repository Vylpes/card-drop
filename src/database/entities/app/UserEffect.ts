import {Column, Entity} from "typeorm";
import AppBaseEntity from "../../../contracts/AppBaseEntity";
import AppDataSource from "../../dataSources/appDataSource";

@Entity()
export default class UserEffect extends AppBaseEntity {
    constructor(name: string, userId: string, unused: number, WhenExpires?: Date) {
        super();

        this.Name = name;
        this.UserId = userId;
        this.Unused = unused;
        this.WhenExpires = WhenExpires;
    }

    @Column()
    Name: string;

    @Column()
    UserId: string;

    @Column()
    Unused: number;

    @Column({ nullable: true })
    WhenExpires?: Date;

    public AddUnused(amount: number) {
        this.Unused += amount;
    }

    public UseEffect(whenExpires: Date): boolean {
        if (this.Unused == 0) {
            return false;
        }

        this.Unused -= 1;
        this.WhenExpires = whenExpires;

        return true;
    }

    public IsEffectActive(): boolean {
        const now = new Date();

        if (this.WhenExpires && now < this.WhenExpires) {
            return true;
        }

        return false;
    }

    public static async FetchOneByUserIdAndName(userId: string, name: string): Promise<UserEffect | null> {
        const repository = AppDataSource.getRepository(UserEffect);

        const single = await repository.findOne({ where: { UserId: userId, Name: name } });

        return single;
    }
}
