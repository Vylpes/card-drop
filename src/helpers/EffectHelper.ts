import {EmbedBuilder} from "discord.js";
import UserEffect from "../database/entities/app/UserEffect";
import EmbedColours from "../constants/EmbedColours";

export default class EffectHelper {
    public static async AddEffectToUserInventory(userId: string, name: string, quantity: number = 1) {
        let effect = await UserEffect.FetchOneByUserIdAndName(userId, name);

        if (!effect) {
            effect = new UserEffect(name, userId, quantity);
        } else {
            effect.AddUnused(quantity);
        }

        await effect.Save(UserEffect, effect);
    }

    public static async UseEffect(userId: string, name: string, whenExpires: Date): Promise<boolean> {
        const effect = await UserEffect.FetchOneByUserIdAndName(userId, name);
        const now = new Date();

        if (!effect || effect.Unused == 0) {
            return false;
        }

        if (effect.WhenExpires && now < effect.WhenExpires) {
            return false;
        }

        effect.UseEffect(whenExpires);

        await effect.Save(UserEffect, effect);

        return true;
    }

    public static async HasEffect(userId: string, name: string): Promise<boolean> {
        const effect = await UserEffect.FetchOneByUserIdAndName(userId, name);
        const now = new Date();

        if (!effect || !effect.WhenExpires) {
            return false;
        }

        if (now > effect.WhenExpires) {
            return false;
        }

        return true;
    }

    public static async GenerateEffectEmbed(userId: string, page: number): Promise<EmbedBuilder> {
        const effects = await UserEffect.FetchAllByUserIdPaginated(userId, page - 1);

        let description = "*none*";

        if (effects.length > 0) {
            description = effects.map(x => `${x.Name} x${x.Unused}`).join("\n");
        }

        const embed = new EmbedBuilder()
            .setTitle("Effects")
            .setDescription(description)
            .setColor(EmbedColours.Ok);

        return embed;
    }
}
