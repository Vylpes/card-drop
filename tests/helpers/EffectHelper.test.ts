import UserEffect from "../../src/database/entities/app/UserEffect";
import EffectHelper from "../../src/helpers/EffectHelper";

describe("AddEffectToUserInventory", () => {
    describe("GIVEN effect is in database", () => {
        const effectMock = {
            AddUnused: jest.fn(),
            Save: jest.fn(),
        };

        beforeAll(async () => {
            UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(effectMock);

            await EffectHelper.AddEffectToUserInventory("userId", "name", 1);
        });

        test("EXPECT database to be fetched", () => {
            expect(UserEffect.FetchOneByUserIdAndName).toHaveBeenCalledTimes(1);
            expect(UserEffect.FetchOneByUserIdAndName).toHaveBeenCalledWith("userId", "name");
        });

        test("EXPECT effect to be updated", () => {
            expect(effectMock.AddUnused).toHaveBeenCalledTimes(1);
            expect(effectMock.AddUnused).toHaveBeenCalledWith(1);
        });

        test("EXPECT effect to be saved", () => {
            expect(effectMock.Save).toHaveBeenCalledTimes(1);
            expect(effectMock.Save).toHaveBeenCalledWith(UserEffect, effectMock);
        });
    });

    describe("GIVEN effect is not in database", () => {
        beforeAll(async () => {
            UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(null);
            UserEffect.prototype.Save = jest.fn();

            await EffectHelper.AddEffectToUserInventory("userId", "name", 1);
        });

        test("EXPECT effect to be saved", () => {
            expect(UserEffect.prototype.Save).toHaveBeenCalledTimes(1);
            expect(UserEffect.prototype.Save).toHaveBeenCalledWith(UserEffect, expect.any(UserEffect));
        });
    });
});

describe("UseEffect", () => {
    describe("GIVEN effect is in database", () => {
        describe("GIVEN now is before effect.WhenExpires", () => {
            let result: boolean | undefined;

            // nowMock < whenExpires
            const nowMock = new Date(2024, 11, 3, 13, 30);
            const whenExpires = new Date(2024, 11, 3, 14, 0);

            const userEffect = {
                Unused: 1,
                WhenExpires: whenExpires,
            };

            beforeAll(async () => {
                jest.setSystemTime(nowMock);

                UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                result = await EffectHelper.UseEffect("userId", "name", new Date());
            });

            test("EXPECT false returned", () => {
                expect(result).toBe(false);
            });
        });

        describe("GIVEN currently used effect is inactive", () => {
            let result: boolean | undefined;

            // nowMock > whenExpires
            const nowMock = new Date(2024, 11, 3, 13, 30);
            const whenExpires = new Date(2024, 11, 3, 13, 0);
            const whenExpiresNew = new Date(2024, 11, 3, 15, 0);

            const userEffect = {
                Unused: 1,
                WhenExpires: whenExpires,
                UseEffect: jest.fn(),
                Save: jest.fn(),
            };

            beforeAll(async () => {
                jest.setSystemTime(nowMock);

                UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                result = await EffectHelper.UseEffect("userId", "name", whenExpiresNew);
            });

            test("EXPECT UseEffect to be called", () => {
                expect(userEffect.UseEffect).toHaveReturnedTimes(1);
                expect(userEffect.UseEffect).toHaveBeenCalledWith(whenExpiresNew);
            });

            test("EXPECT effect to be saved", () => {
                expect(userEffect.Save).toHaveBeenCalledTimes(1);
                expect(userEffect.Save).toHaveBeenCalledWith(UserEffect, userEffect);
            });

            test("EXPECT true returned", () => {
                expect(result).toBe(true);
            });
        });

        describe("GIVEN effect.WhenExpires is null", () => {
            let result: boolean | undefined;

            // nowMock > whenExpires
            const nowMock = new Date(2024, 11, 3, 13, 30);
            const whenExpiresNew = new Date(2024, 11, 3, 15, 0);

            const userEffect = {
                Unused: 1,
                WhenExpires: null,
                UseEffect: jest.fn(),
                Save: jest.fn(),
            };

            beforeAll(async () => {
                jest.setSystemTime(nowMock);

                UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                result = await EffectHelper.UseEffect("userId", "name", whenExpiresNew);
            });

            test("EXPECT UseEffect to be called", () => {
                expect(userEffect.UseEffect).toHaveBeenCalledTimes(1);
                expect(userEffect.UseEffect).toHaveBeenCalledWith(whenExpiresNew);
            });

            test("EXPECT effect to be saved", () => {
                expect(userEffect.Save).toHaveBeenCalledTimes(1);
                expect(userEffect.Save).toHaveBeenCalledWith(UserEffect, userEffect);
            });

            test("EXPECT true returned", () => {
                expect(result).toBe(true);
            });
        });
    });

    describe("GIVEN effect is not in database", () => {
        let result: boolean | undefined;

        // nowMock > whenExpires
        const nowMock = new Date(2024, 11, 3, 13, 30);
        const whenExpiresNew = new Date(2024, 11, 3, 15, 0);

        beforeAll(async () => {
            jest.setSystemTime(nowMock);

            UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(null);

            result = await EffectHelper.UseEffect("userId", "name", whenExpiresNew);
        });

        test("EXPECT false returned", () => {
            expect(result).toBe(false);
        });
    });

    describe("GIVEN effect.Unused is 0", () => {
        let result: boolean | undefined;

        // nowMock > whenExpires
        const nowMock = new Date(2024, 11, 3, 13, 30);
        const whenExpiresNew = new Date(2024, 11, 3, 15, 0);

        const userEffect = {
            Unused: 0,
            WhenExpires: null,
            UseEffect: jest.fn(),
            Save: jest.fn(),
        };

        beforeAll(async () => {
            jest.setSystemTime(nowMock);

            UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

            result = await EffectHelper.UseEffect("userId", "name", whenExpiresNew);
        });

        test("EXPECT false returned", () => {
            expect(result).toBe(false);
        });
    });
});

describe("HasEffect", () => {
    describe("GIVEN effect is in database", () => {
        describe("GIVEN effect.WhenExpires is defined", () => {
            describe("GIVEN now is before effect.WhenExpires", () => {
                let result: boolean | undefined;

                const nowMock = new Date(2024, 11, 3, 13, 30);
                const whenExpires = new Date(2024, 11, 3, 15, 0);

                const userEffect = {
                    WhenExpires: whenExpires,
                };

                beforeAll(async () => {
                    jest.setSystemTime(nowMock);

                    UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                    result = await EffectHelper.HasEffect("userId", "name");
                });

                test("EXPECT true returned", () => {
                    expect(result).toBe(true);
                });
            });

            describe("GIVEN now is after effect.WhenExpires", () => {
                let result: boolean | undefined;

                const nowMock = new Date(2024, 11, 3, 16, 30);
                const whenExpires = new Date(2024, 11, 3, 15, 0);

                const userEffect = {
                    WhenExpires: whenExpires,
                };

                beforeAll(async () => {
                    jest.setSystemTime(nowMock);

                    UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                    result = await EffectHelper.HasEffect("userId", "name");
                });

                test("EXPECT false returned", () => {
                    expect(result).toBe(false);
                });
            });
        });

        describe("GIVEN effect.WhenExpires is undefined", () => {
            let result: boolean | undefined;

            const userEffect = {
                WhenExpires: undefined,
            };

            beforeAll(async () => {
                UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(userEffect);

                result = await EffectHelper.HasEffect("userId", "name");
            });

            test("EXPECT false returned", () => {
                expect(result).toBe(false);
            });
        });
    });

    describe("GIVEN effect is not in database", () => {
        let result: boolean | undefined;

        beforeAll(async () => {
            UserEffect.FetchOneByUserIdAndName = jest.fn().mockResolvedValue(null);

            result = await EffectHelper.HasEffect("userId", "name");
        });

        test("EXPECT false returned", () => {
            expect(result).toBe(false);
        });
    });
});
