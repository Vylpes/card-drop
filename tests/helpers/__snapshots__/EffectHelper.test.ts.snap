// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateEffectEmbed GIVEN user does NOT have an effect, EXPECT empty embed to be returned 1`] = `
{
  "embed": {
    "color": 3166394,
    "description": "*none*",
    "footer": {
      "icon_url": undefined,
      "text": "Page 1 of 1",
    },
    "title": "Effects",
  },
  "row": {
    "components": [
      {
        "custom_id": "effects list 0",
        "disabled": true,
        "emoji": undefined,
        "label": "Previous",
        "style": 1,
        "type": 2,
      },
      {
        "custom_id": "effects list 2",
        "disabled": true,
        "emoji": undefined,
        "label": "Next",
        "style": 1,
        "type": 2,
      },
    ],
    "type": 1,
  },
}
`;

exports[`GenerateEffectEmbed GIVEN user has an effect, EXPECT detailed embed to be returned 1`] = `
{
  "embed": {
    "color": 3166394,
    "description": "Unclaimed Chance Up x1",
    "footer": {
      "icon_url": undefined,
      "text": "Page 1 of 1",
    },
    "title": "Effects",
  },
  "row": {
    "components": [
      {
        "custom_id": "effects list 0",
        "disabled": true,
        "emoji": undefined,
        "label": "Previous",
        "style": 1,
        "type": 2,
      },
      {
        "custom_id": "effects list 2",
        "disabled": true,
        "emoji": undefined,
        "label": "Next",
        "style": 1,
        "type": 2,
      },
    ],
    "type": 1,
  },
}
`;

exports[`GenerateEffectEmbed GIVEN user has more than 1 page of effects, EXPECT pagination enabled 1`] = `
{
  "embed": {
    "color": 3166394,
    "description": "Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1",
    "footer": {
      "icon_url": undefined,
      "text": "Page 1 of 2",
    },
    "title": "Effects",
  },
  "row": {
    "components": [
      {
        "custom_id": "effects list 0",
        "disabled": true,
        "emoji": undefined,
        "label": "Previous",
        "style": 1,
        "type": 2,
      },
      {
        "custom_id": "effects list 2",
        "disabled": false,
        "emoji": undefined,
        "label": "Next",
        "style": 1,
        "type": 2,
      },
    ],
    "type": 1,
  },
}
`;

exports[`GenerateEffectEmbed GIVEN user is on a page other than 1, EXPECT pagination enabled 1`] = `
{
  "embed": {
    "color": 3166394,
    "description": "Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1
Unclaimed Chance Up x1",
    "footer": {
      "icon_url": undefined,
      "text": "Page 2 of 2",
    },
    "title": "Effects",
  },
  "row": {
    "components": [
      {
        "custom_id": "effects list 1",
        "disabled": false,
        "emoji": undefined,
        "label": "Previous",
        "style": 1,
        "type": 2,
      },
      {
        "custom_id": "effects list 3",
        "disabled": true,
        "emoji": undefined,
        "label": "Next",
        "style": 1,
        "type": 2,
      },
    ],
    "type": 1,
  },
}
`;
